### PR TITLE
feat(MMST): multi minimum spanning tree environment

### LINKED ISSUE #128: feat(cmst): full environment
Implement the cooperative minimum spanning tree environment.

### PR DESCRIPTION
**Pull request by ulricharmel**
Full implementation with docs and actor critic network.

[Here](https://app.neptune.ai/o/InstaDeep/org/jumanji/runs/compare?viewId=standard-view&detailsTab=charts&shortId=JUJU-972&dash=charts&type=run&compare=IwBgzAHANMQ) is the link with the run comparing the actor-critc and the random agent for the defaults environment configs.

Close #128

### CODE REVIEW

File: jumanji/environments/routing/cmst/env.py [hunk 1]
```diff
@@ -0,0 +1,749 @@
...
+        num_nodes: int = 12,
+        num_edges: int = 24,
+        max_degree: int = 5,
+        num_agents: int = 2,
+        num_nodes_per_agent: int = 3,
+        step_limit: int = 70,
+        reward_fn: Optional[RewardFn] = None,
+        reward_values: Tuple = (0.1, -0.03, -0.01),  # (5.0, -0.2, -0.1)
+        generator_fn: Type[Generator] = SplitRandomGenerator,
+        renderer: Optional[Renderer] = None,
```
Comment (clement-bonnet) [inline jumanji/environments/routing/cmst/env.py:124]: Can we please call `renderer` a `viewer` instead? For consistency. Similarly, `Renderer` -> `Viewer`.

File: jumanji/environments/routing/cmst/env.py [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+            - INVALID_ALREADY_TRAVERSED = -3
+    """
+
+    def __init__(
+        self,
+        num_nodes: int = 12,
+        num_edges: int = 24,
+        max_degree: int = 5,
+        num_agents: int = 2,
+        num_nodes_per_agent: int = 3,
```
Comment (clement-bonnet) [inline jumanji/environments/routing/cmst/env.py:119]: Can we please group everything related to the generator into the generator object? Please see how it is done in `Connector` for reference. The `__init__` method should only take as arguments: `generator`, `step_limit`, `reward_fn`, `viewer`. Thanks!

File: requirements/requirements.txt [hunk 3]
```diff
@@ -4,6 +4,7 @@ gym>=0.22.0
 jax>=0.2.26
 jaxlib>=0.1.74
 matplotlib>=3.3.4
+networkx==2.8.3
```
Comment (clement-bonnet) [inline requirements/requirements.txt:7]: While I understand `networkx` may come handy for graph rendering, we are trying to keep Jumanji as light as possible. Is there any way to implement rendering using Matplotlib instead? `GraphColoring` has rendering implemented using Matplotlib and most features should be transferable to CMST. Could you check with [Elsh](ElshadaiK)?

File: jumanji/environments/routing/cmst/viewer.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+
+        self._name = name
+        self.num_agents = num_agents
+
+        # Pick display method. Only one mode avaliable
+        self._display: Callable[[plt.Figure], Optional[NDArray]]
+        self._display = self._display_human
+
+        np.random.seed(0)
+        random.seed(0)
```
Comment (clement-bonnet) [inline jumanji/environments/routing/cmst/viewer.py:59]: I don't think the `random` package from the standard library is needed here.

File: jumanji/environments/routing/cmst/viewer.py [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+grey = (100 / 255, 100 / 255, 100 / 255)
+white = (255 / 255, 255 / 255, 255 / 255)
+yellow = (200 / 255, 200 / 255, 0 / 255)
+red = (200 / 255, 0 / 255, 0 / 255)
+black = (0 / 255, 0 / 255, 0 / 255)
+blue = (50 / 255, 50 / 255, 160 / 255)
+
+
+class Renderer(Viewer):
```
Comment (clement-bonnet) [inline jumanji/environments/routing/cmst/viewer.py:37]: Can we call this `CMSTViewer` instead?

File: jumanji/environments/routing/cmst/reward.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        Args:
+            state: Environment state
+            actions: Actions taken by all the agents to reach this state.
+            nodes_to_connect: Array containing the nodes each agent needs to connect.
+        Returns:
+            reward
+        """
+
+
+class DefaultRewardFn(RewardFn):
```
Comment (clement-bonnet) [inline jumanji/environments/routing/cmst/reward.py:46]: Could we give it a more explicit name? The fact that it is the default used by the environment does not characterise the reward function in itself.

File: jumanji/environments/routing/cmst/reward.py
Comment (clement-bonnet) [inline jumanji/environments/routing/cmst/reward.py:1]: Would it be possible to implement a quick `reward_test.py` for the implemented reward function?

**Review by clement-bonnet — CHANGES_REQUESTED** [review]
Thanks so much for your contribution! It is a difficult environment. I have left a few preliminary comments before digging deeper into the code. Please let me know if they sound reasonable.
Thanks!

File: jumanji/environments/routing/cmst/env.py [hunk 7]
```diff
@@ -0,0 +1,749 @@
...
+            num_nodes_per_agent,
+            step_limit,
+        )
+
+        self._renderer = renderer
+
+    def __repr__(self) -> str:
+        return (
+            f"CMST(num_nodes={self.num_nodes}, num_edges={self.num_edges}, "
+            f"num_agents={self.num_agents}, num_components={self.num_nodes_per_agent})"
```
[the remark this answers is not in the document]
Comment (ulricharmel) [reply jumanji/environments/routing/cmst/env.py]: I have actually deleted this function as I see we don't have it anymore in the connector environment.

File: jumanji/environments/routing/cmst/env.py [hunk 8]
```diff
@@ -0,0 +1,749 @@
...
+    def __init__(
+        self,
+        num_nodes: int = 12,
+        num_edges: int = 24,
+        max_degree: int = 5,
+        num_agents: int = 2,
+        num_nodes_per_agent: int = 3,
+        step_limit: int = 70,
+        reward_fn: Optional[RewardFn] = None,
+        reward_values: Tuple = (0.1, -0.03, -0.01),  # (5.0, -0.2, -0.1)
```
[the remark this answers is not in the document]
Comment (ulricharmel) [reply jumanji/environments/routing/cmst/env.py]: This has been fixed

File: jumanji/environments/routing/cmst/env.py (hunk 1 above)
Comment (ulricharmel) [reply jumanji/environments/routing/cmst/env.py]: This has been fixed.

File: jumanji/environments/routing/cmst/env.py (hunk 2 above)
Comment (ulricharmel) [reply jumanji/environments/routing/cmst/env.py]: I have fixed this.

File: requirements/requirements.txt (hunk 3 above)
Comment (ulricharmel) [reply requirements/requirements.txt]: I have removed the need for networkx. I wasn't using it for something important.

File: jumanji/environments/routing/cmst/viewer.py (hunk 4 above)
Comment (ulricharmel) [reply jumanji/environments/routing/cmst/viewer.py]: This has been fixed

File: jumanji/environments/routing/cmst/viewer.py (hunk 5 above)
Comment (ulricharmel) [reply jumanji/environments/routing/cmst/viewer.py]: Fixed.

File: jumanji/environments/routing/cmst/reward.py (hunk 6 above)
Comment (ulricharmel) [reply jumanji/environments/routing/cmst/reward.py]: I have named it DenseRewardFn

File: jumanji/environments/routing/cmst/reward.py
Comment (ulricharmel) [reply jumanji/environments/routing/cmst/reward.py:1]: I actually had a test for this in the env_test.py but I have moved that to the reward_test file.

File: jumanji/environments/routing/mmst/types.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    adj_matrix: chex.Array  # (num_nodes, num_nodes)
+    connected_nodes: chex.Array  # (num_agents, step_limit)
+    connected_nodes_index: chex.Array  # (num_agents, num_nodes)
+    nodes_to_connect: chex.Array  # (num_agents, num_nodes_to_connect_per_agent)
+    node_edges: chex.Array  # (num_agents, num_nodes, num_nodes)
+    positions: chex.Array  # (num_agents,)
+    position_index: chex.Array  # (num_agents,)
+    action_mask: chex.Array  # (num_agents, num_nodes)
+    finished_agents: chex.Array  # (num_agents,)
+    step_count: jnp.int32
```
Comment (ElshadaiK) [inline jumanji/environments/routing/mmst/types.py:53]:
```suggestion
    step_count: jnp.int32  # ()
```

File: jumanji/environments/routing/mmst/types.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    key: chex.PRNGKey
```
Comment (ElshadaiK) [inline jumanji/environments/routing/mmst/types.py:54]:
```suggestion
    key: chex.PRNGKey  # (2,)
```

File: jumanji/environments/routing/mmst/types.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+class Graph:
+    """
+    nodes: Array with node indices (jnp.arange(number of nodes)).
+    edges: Array with all egdes in the graph.
+    edge_codes: Array with edge codes.
+    max_degree: (int).
+    node_degree: Array with the degree of every node.
+    edge_index: (int) index location for the next edge.
+    """
+
```
Comment (ElshadaiK) [inline jumanji/environments/routing/mmst/types.py:111]: Providing the shapes would help

File: jumanji/environments/routing/mmst/env.py [hunk 12]
```diff
@@ -0,0 +1,721 @@
...
+
+class MMST(Environment[State]):
+    """The `MMST` (Multi Minimum Spanning Tree) environment
+    consists of a random connected graph
+    with groups of nodes (same node types) that needs to be connected.
+    The goal of the environment is to connect all nodes of the same type together
+    without using the same utility nodes (nodes that do not belong to any group of nodes).
+
+    Note: routing problems are randomly generated and may not be solvable!
+    Additionally, the total number of nodes should be at least 20% more than
```
Comment (ElshadaiK) [inline jumanji/environments/routing/mmst/env.py:48]:
```suggestion
    Requirements:
    The total number of nodes should be at least 20% more than
```

File: jumanji/environments/routing/mmst/env.py [hunk 13]
```diff
@@ -0,0 +1,721 @@
...
+        self,
+        generator_fn: Optional[Generator] = None,
+        reward_fn: Optional[RewardFn] = None,
+        step_limit: int = 100,
+        viewer: Optional[Viewer[State]] = None,
+    ):
+        """Create the `MMST` environment.
+
+        Args:
+            generator_fn: `Generator` whose `__call__` instantiates an environment instance.
```
Comment (ElshadaiK) [inline jumanji/environments/routing/mmst/env.py:125]: It would be easily understandable if we have the parameters the default `generator_fn` is inistantiated with in the docstring

File: jumanji/environments/routing/mmst/env.py [hunk 14]
```diff
@@ -0,0 +1,721 @@
...
+            max_step=step_limit,
+        )
+
+        self.num_agents = self._generator_fn.num_agents
+        self.num_nodes = self._generator_fn.num_nodes
+        self.num_nodes_per_agent = self._generator_fn.num_nodes_per_agent
+
+        self._reward_fn = reward_fn or DenseRewardFn(reward_values=(1.0, -0.03, -0.03))
+
+        self._renderer = viewer
```
Comment (ElshadaiK) [inline jumanji/environments/routing/mmst/env.py:148]: naming this variable `_viewer` or `_env_viewer` would be better than renderer. We're also animating using the viewer

File: jumanji/environments/routing/mmst/env.py [hunk 15]
```diff
@@ -0,0 +1,721 @@
...
+            num_values=jnp.full((self.num_agents,), self.num_nodes, jnp.int32),
+            name="action",
+        )
+
+    def observation_spec(self) -> specs.Spec[Observation]:
+        """Returns the observation spec.
+
+        Returns:
+            observation_spec: a Tuple containing the spec for each of the constituent fields of an
+            observation.
```
Comment (ElshadaiK) [inline jumanji/environments/routing/mmst/env.py:167]: Can you also list the fields for the `Observation`?

File: jumanji/environments/routing/mmst/generator.py [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+        self._max_degree = max_degree
+        self._num_agents = num_agents
+        self._num_nodes_per_agent = num_nodes_per_agent
+        self._total_comps = num_nodes_per_agent * num_agents
+        self._max_step = max_step
+
+        if num_nodes_per_agent * num_agents > num_nodes * 0.8:
+            raise ValueError(
+                f"The number of nodes to connect i.e. {num_nodes_per_agent * num_agents} "
+                f"should be much less than the number of nodes, which is {int(0.8*num_nodes)}."
```
Comment (ElshadaiK) [inline jumanji/environments/routing/mmst/generator.py:61]:
```suggestion
                f"should be much less than the number of nodes, which is {int(0.8*num_nodes)}."
                f"This is to guarantee there are enough remaining nodes to "
                f"create a path with all the nodes we want to connect."
```

File: jumanji/environments/routing/mmst/env.py [hunk 17]
```diff
@@ -0,0 +1,762 @@
...
+            (added_nodes, new_actions, action, nodes, agent_permutation, 0),
+        )
+
+        def mask_visited_nodes(
+            node_visited: jnp.int32, oldaction: jnp.int32
+        ) -> jnp.int32:
+            new_action = jax.lax.cond(  # type:ignore
+                node_visited != EMPTY_NODE,
+                lambda *_: INVALID_ALREADY_TRAVERSED,
+                lambda *_: oldaction,
```
Comment (cemlyn007) [inline jumanji/environments/routing/mmst/env.py:638]:
```suggestion
                lambda *_: old_action,
```

File: jumanji/environments/routing/mmst/env.py [hunk 18]
```diff
@@ -0,0 +1,762 @@
...
+            agent_permutation,
+            index,
+        ) = jax.lax.while_loop(
+            not_all_agents_actions_examined,
+            modify_action_if_agent_target_node_is_selected,
+            (added_nodes, new_actions, action, nodes, agent_permutation, 0),
+        )
+
+        def mask_visited_nodes(
+            node_visited: jnp.int32, oldaction: jnp.int32
```
Comment (cemlyn007) [inline jumanji/environments/routing/mmst/env.py:633]:
```suggestion
            node_visited: jnp.int32, old_action: jnp.int32
```

File: jumanji/environments/routing/mmst/env.py [hunk 19]
```diff
@@ -0,0 +1,762 @@
...
+        state = State(
+            node_types=node_types,
+            adj_matrix=adj_matrix,
+            nodes_to_connect=nodes_to_connect,
+            connected_nodes=conn_nodes,
+            connected_nodes_index=conn_nodes_index,
+            position_index=jnp.zeros((self.num_agents), dtype=jnp.int32),
+            positions=agents_pos,
+            node_edges=active_node_edges,
+            action_mask=self._update_action_mask(
```
[the remark this answers is not in the document]
Comment (ulricharmel) [reply jumanji/environments/routing/mmst/env.py]: Thanks @cemlyn007 I have addressed your comments.

File: jumanji/environments/routing/mmst/env.py [hunk 20]
```diff
@@ -0,0 +1,762 @@
...
+            return new_action
+
+        final_actions = jnp.zeros_like(new_actions)
+        # set the action to 0 if the agent is moving to an already connected node
+        for agent in range(self.num_agents):
+            node_visited = state.connected_nodes_index[agent, nodes[agent]]
+            new_action = mask_visited_nodes(node_visited, new_actions[agent])
+            final_actions = final_actions.at[agent].set(new_action)
+
+        # masked agents with finished states
```
Comment (cemlyn007) [inline jumanji/environments/routing/mmst/env.py:632]:
```suggestion
        # Masked agents with finished states.
```
Please can you review and make sure comments are proper sentences?

Comment (ulricharmel) [reply jumanji/environments/routing/mmst/env.py:632]: Sure I will go through all the comments in the environment now.

**Review by clement-bonnet — CHANGES_REQUESTED** [review]
There are still a few inconsistencies. Would you mind fixing these before we merge?

File: jumanji/environments/routing/mmst/generator.py [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+        """Initialises arrays to hold environment states"""
+
+        state_nodes_to_connect = EMPTY_NODE * (
+            jnp.ones((self._num_agents, self._num_nodes_per_agent), dtype=jnp.int32)
+        )
+
+        node_types = EMPTY_NODE * jnp.ones(self._num_nodes, dtype=jnp.int32)
+        conn_nodes = EMPTY_NODE * jnp.ones(
+            (self._num_agents, self._max_step), dtype=jnp.int32
+        )
```
[the remark this answers is not in the document]
Comment (ulricharmel) [reply jumanji/environments/routing/mmst/generator.py:198]: This is to initialisation the connection path followed by an agent. So we initialise this array to have size equal to the maximum number of steps.

File: jumanji/environments/routing/mmst/utils.py [hunk 22]
```diff
@@ -33,6 +37,80 @@ def build_adjecency_matrix(num_nodes: int, edges: jnp.ndarray) -> jnp.ndarray:
...
+    """
+
+    full_action_mask = node_edges != EMPTY_NODE
+    action_mask = jnp.zeros((num_agents, num_nodes), dtype=bool)
+    for agent in range(num_agents):
+        node = position[agent]
+        action_mask = action_mask.at[agent].set(
+            full_action_mask[agent, node] * ~finished_agents[agent]
+        )
+
```
Comment (clement-bonnet) [inline jumanji/environments/routing/mmst/utils.py:110]: Why doing a sequential for loop here? Can we parallelize the process? E.g. does the following work?
```suggestion
    action_mask = full_action_mask[jnp.arange(num_agents), position] & ~finished_agents
```

**Review by clement-bonnet — APPROVED** [review]
Looks good to me!
